src/eh1/io.rs

```diff
@@ -0,0 +1,617 @@
+//! IO mock implementations.
+//!
+//! This mock supports the specification and checking of expectations to allow
+//! automated testing of IO based drivers. Mismatches between expected and
+//! real IO transactions will cause runtime assertions to assist with locating
+//! faults.
+//!
+//! ## Usage
+//!
+//! ```
+//! use embedded_io::{self, Read, Write};
+//! // optional async
+//! use embedded_io_async;
+//! use embedded_hal_mock::eh1::io::{Mock as IoMock, Transaction as IoTransaction};
+//!
+//!
+//!
+//! // Configure expectations
+//! let expectations = [
+//!     IoTransaction::write(vec![10, 20]),
+//!     IoTransaction::read(vec![20, 30]),
+//!     IoTransaction::flush(),
+//! ];
+//!
+//! let mut io = IoMock::new(&expectations);
+//!
+//! // Writing
+//! let ret = io.write(&[10, 20]).unwrap();
+//! assert_eq!(ret, 2);
+//!
+//! // Reading
+//! let mut buffer = [0; 2];
+//! let ret = io.read(&mut buffer).unwrap();
+//! assert_eq!(buffer, [20, 30]);
+//! assert_eq!(ret, 2);
+//!
+//! // Flushing
+//! io.flush().unwrap();
+//!
+//! // Also async
+//! async {
+//!     let ret = embedded_io_async::Write::write(&mut io, &[10, 20]).await.unwrap();
+//! };
+//!
+//! // Finalizing expectations
+//! io.done();
+//! ```
+
+use embedded_io::{
+    self, BufRead, ErrorKind, ErrorType, Read, ReadReady, Seek, SeekFrom, Write, WriteReady,
+};
+
+use crate::common::Generic;
+
+/// IO Transaction mode
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Mode {
+    /// Write transaction
+    Write,
+    /// Read transaction
+    Read,
+    /// Flush transaction
+    Flush,
+    /// Seek transaction
+    Seek(SeekFrom),
+    /// WriteReady transaction
+    WriteReady(bool),
+    /// ReadReady transaction
+    ReadReady(bool),
+    /// FillBuff transaction
+    FillBuff,
+    /// Consume transaction
+    Consume(usize),
+}
```

Comment:
As a thought: Might it be better for readability if the variants with associated data (e.g. `WriteReady(bool)`) would use named parameters (e.g. `WriteReady { ready: bool }`)?

I have no strong opinon on this, so if you prefer the brevity, that's fine too. (After all, this type is usually instantiated through the constructor methods of the transaction anyways.)

Reply:
Fine for me.